build-util/src/bindgen.rs

```diff
@@ -0,0 +1,197 @@
+use std::{env, fs, io, path::Path, process};
+
+use quote::ToTokens;
+use syn::{ForeignItem, Item, Type};
+
+use crate::link_visitor::{
+    syn::{self, visit_mut::VisitMut},
+    Link,
+};
+
+pub fn generate(
+    vita_headers_include: &Path,
+    db: &Path,
+    bindings_output: &Path,
+    vitasdk_sys_manifest: &Path,
+    is_build_rs: bool,
+) {
+    log::info!("Generating preprocessed bindings");
+    let headers = vitasdk_sys_manifest.parent().unwrap().join("headers");
+    let bindings = generate_preprocessed_bindings(&headers, vita_headers_include, is_build_rs);
+
+    let mut bindings = syn::parse_file(&bindings).expect("failed to parse file");
+
+    log::info!(
+        "Loading vita-headers metadata yaml files from \"{}\"",
+        db.to_string_lossy()
+    );
+    let mut link = Link::load(db, vitasdk_sys_manifest);
+    link.visit_file_mut(&mut bindings);
+
+    // We sort items here so generated bindings don't depend on the included order.
+    sort_items(&mut bindings.items);
+
+    let bindings = bindings.into_token_stream();
+
+    {
+        log::info!(
+            "Writing postprocessed bindings into {}",
+            bindings_output.to_string_lossy()
+        );
+        let mut bindings_output = io::BufWriter::new(fs::File::create(bindings_output).unwrap());
+        use std::io::Write;
+        write!(bindings_output, "{bindings}").unwrap();
+    }
+
+    let cargo = env::var_os("CARGO");
+    let mut fmt_cmd = process::Command::new(cargo.as_deref().unwrap_or_else(|| "cargo".as_ref()));
+    fmt_cmd.args(["fmt", "--"]);
+    fmt_cmd.arg(bindings_output);
+
+    log::info!("Running formatting command: {fmt_cmd:?}");
+    let exit_status = fmt_cmd.status().unwrap();
+    if exit_status.success() {
+        log::info!("Formatting command finished");
+    } else {
+        log::warn!("Formatting command failed with status: {exit_status:?}");
```

Comment:
We should panic on failure if generating `src/bindings.rs`, since we want the code to be stable.

Reply:
Done!